contributors/design-proposals/dynamic-admission-control-configuration.md

```diff
@@ -0,0 +1,243 @@
+    // ports open, port 443 will be used if it is open, otherwise it is an error.
+	Service ServiceReference
+	// CABundle is a PEM encoded CA bundle which will be used to validate webhook's server certificate.
+	CABundle []byte
+}
+
+// ServiceReference holds a reference to Service.legacy.k8s.io
+type ServiceReference struct {
+	// Namespace is the namespace of the service
+	Namespace string
+	// Name is the name of the service
+	Name string
+}
+```
+
+## Synchronization of AdmissionControlConfiguration 
+
+If the `initializer admission controller` and the `generic webhook admission
+controller` watch the `AdmissionControlConfiguration` and act upon deltas, their
+cached version of the configuration might be arbitrarily delayed. This makes it
+impossible to predict what initializer/hooks will be applied to newly created
+objects.
+
+We propose the following way to make the behavior of the `initializer admission
+controller` and the `generic webhook admission controller` predictable.
+
+#### 1. Do consistent read of AdmissionControlConfiguration periodically
+
+The `initializer admission controller` and the `generic webhook admission
+controller` do a consistent read of the AdmissionControlConfiguration every 1s.
+If there isn't any successful read in the last 5s, the two admission controllers 
+block all incoming request.
+One consistent read per second isn't going to cause performance issues.
+
+## What if an initializer controller/webhook is not ready after registered?
+
+This will block the entire cluster. We have a few options:
+
+1. only allow initializers/webhooks to be created as "fail open". This could be
+   enforced via validation. They can upgrade themselves to "fail closed" via the
+   normal Update operation. A human can also update them to "fail closed" later. 
+
+2. less preferred: add readiness check to initializer and webhooks, `initializer
+   admission controller` and `generic webhook admission controller` only apply
+   those have passed readiness check. Specifically, we add `readiness` fields to
+   `AdmissionControllerConfiguration`; then we either create yet another
+   controller to probe for the readiness and update the
+   `AdmissionControllerConfiguration`, or ask each initializer/webhook to update
+   their readiness in the `AdmissionControllerConfigure`. The former is complex.
+   The latter is essentially the same as the first approach, except that we need
+   to introduce the additional concept of "readiness".
+
+## Handling fail-open initializers
+
+#132 assumed initializers always failed closed. It is dangerous since crashed 
+initializers can block the whole cluster. We propose to allow initializers to 
+fail open, and in 1.7, let all initializers fail open.
+
+A `fail-open initializers controller` will remove the timed out fail-open
+initializers from objects' initializers list. Every 30s, the controller 
+
+* lists uninitialized objects
```

Comment:
We'd want to use the shared informers, and we only need to start watching things that have an initializer.  That causes a problem though, because we may have to start watching something that has an initializer.

I don't want to list all every 30s though, that is definitely too heavy.  I'm not sure how often this really has to happen.  I kind of would prefer this be more read repair than controller driven, since read repair is very cheap.